src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ArrayBuffer.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Net.Http
+{
+    // Warning: Mutable struct!
+    // The purpose of this struct is to simplify buffer management.
+    // It manages a "read position" and a "write position" into a single array buffer,
+    // and provides some support for resizing/shifting the buffer when necessary.
+
+    // Invariants:
+    // 0 <= _readPosition <= _writePosition <= bytes.Length
+    // Bytes between _writePosition and bytes.Length are available to write to.
+    // Bytes between _readPosition and _writePosition are available to read from.
+    // Bytes between 0 and _readPosition are dead, and should be ignored.
+    internal struct ArrayBuffer
+    {
+        private byte[] _bytes;
+        private int _readPosition;
+        private int _writePosition;
+
+        public ArrayBuffer(int initialSize)
+        {
+            _bytes = new byte[initialSize];
+
+            _readPosition = 0;
+            _writePosition = 0;
+        }
+
+        public Span<byte> ReadBytes => new Span<byte>(_bytes, _readPosition, _writePosition - _readPosition);
+        public Span<byte> WriteBytes => new Span<byte>(_bytes, _writePosition, _bytes.Length - _writePosition);
```

Comment:
Nit: maybe ReadSpan/WriteSpan to correspond to ReadMemory/WriteMemory?

Reply:
ok